apps/pkeyutl.c

```diff
@@ -574,3 +640,67 @@ static int do_keyop(EVP_PKEY_CTX *ctx, int pkey_op,
     }
     return rv;
 }
+
+#define TBUF_MAXSIZE 2048
+
+static int do_raw_keyop(int pkey_op, EVP_PKEY_CTX *ctx,
+                        const EVP_MD *md, EVP_PKEY *pkey, BIO *in,
+                        unsigned char *sig, int siglen,
+                        unsigned char *out, size_t *poutlen)
+{
+    int rv = 0;
+    EVP_MD_CTX *mctx = NULL;
+    unsigned char tbuf[TBUF_MAXSIZE];
+    int tbuf_len = 0;
+
+    if ((mctx = EVP_MD_CTX_new()) == NULL) {
+        BIO_printf(bio_err, "Error: out of memory\n");
+        return rv;
+    }
+    EVP_MD_CTX_set_pkey_ctx(mctx, ctx);
+
+    switch(pkey_op) {
```

Comment:
Can we get here with pkey_op set to anything other than EVP_PKEY_OP_VERIFY or EVP_PKEY_OP_SIGN?

Reply:
I don't think other pkey operation types should go in to this function. There are 5 types that `pkeyutl` can currently handle - sign/verify, encrypt/decrypt and key derivation. It seems public key algorithms are not commonly used to encrypt/decrypt large data directly, so I assume there might be no need to change other operations to stream-reading mode.

Reply:
Right. But should we prevent this function from being called if the op isn't verify or sign? At the moment if the operation was (say) EVP_PKEY_OP_ENCRYPT then wouldn't we get to this function, and then the operation is just silently ignored because it doesn't appear in the switch.

Reply:
So you suggested to add a check at the beginning of this function on the op type?

Reply:
I'd suggest putting a check early on along with the other command line option sanity checks to make sure "-rawin" has only been supplied with "-sign" or "-verify" and display a sensible error message if not. At the moment what happens is `do_raw_keyop` just doesn't do anything and gets to the end of the function returning rv == 0, which ultimately results in the generic error message "Public Key operation error" - which isn't very helpful.

Reply:
Sure, I got your idea, will fix.